PixelHunter1995/SceneLib/SceneParser.cs

```diff
@@ -36,24 +34,90 @@ public static Scene ParseSceneXml(string sceneXmlPath)
                     background = new Background(imagePath, width, height);
                 }
+                else if (node.Name == "tileset")
+                {
+                    string tilesetXmlPathRelative = node.Attributes["source"].Value;
+                    string tilesetXmlPath = Path.Combine(Path.GetDirectoryName(sceneXmlPath),
+                                                         Path.GetDirectoryName(tilesetXmlPathRelative),
+                                                         Path.GetFileName(tilesetXmlPathRelative));
+                    int tilesetFirstGid = int.Parse(node.Attributes["firstgid"].Value);
+                    tileset = ParseTilesetXml(tilesetXmlPath, tilesetFirstGid);
+                }
                 else if (node.Name == "objectgroup" && node.Attributes["name"]?.InnerText == "dogs")
                 {
                     dogs = new List<Dog>();
                     foreach (XmlNode dogNode in node.ChildNodes)
                     {
-                        float x = float.Parse(dogNode.Attributes["x"].Value);
-                        float y = float.Parse(dogNode.Attributes["y"].Value);
-                        float width = float.Parse(dogNode.Attributes["width"].Value);
-                        float height = float.Parse(dogNode.Attributes["height"].Value);
-                        dogs.Add((x, y, width, height));
+                        int x = (int)Math.Round(float.Parse(dogNode.Attributes["x"].Value));
+                        int y = (int)Math.Round(float.Parse(dogNode.Attributes["y"].Value));
+                        int width = (int)Math.Round(float.Parse(dogNode.Attributes["width"].Value));
+                        int height = (int)Math.Round(float.Parse(dogNode.Attributes["height"].Value));
+                        int gid = int.Parse(dogNode.Attributes["gid"].Value);
+                        y = y - height; // Compensate for Tiled's coordinate system
+                        dogs.Add(new Dog(x, y, width, height, gid));
                     }
                 }
                 else if (node.Name == "objectgroup" && node.Attributes["name"]?.InnerText == "walking")
                 {
                     walkingArea = ParseWalkingXml(node);
                 }
             }
-            return new Scene(background, dogs, walkingArea);
+            return new Scene(background, dogs, walkingArea, tileset);
+        }
+
+        private static Tileset ParseTilesetXml(string tilesetXmlPath, int firstGid)
+        {
+            string imagePath = "";
+            int imageWidth = 0;
+            int imageHeight = 0;
+            string name = "";
+            int tileWidth = 0;
+            int tileHeight = 0;
+            int tileCount = 0;
+            int noOfColumns = 0;
+
+            XmlDocument doc = new XmlDocument();
+            doc.Load(tilesetXmlPath);
+
+            foreach (XmlAttribute attribute in doc.DocumentElement.Attributes)
```

Comment:
Why loop over the attributes and not just grab them immediately as above with Dogs and the rest?

Reply:
Good question...